### PR TITLE
Introduce publishMessage(String) in TestReporter API

### LINKED ISSUE #1515: Clarify how TestReporter's keys and values are to be used
About `TestReporter`, the documentation says:

> In JUnit Jupiter you should use `TestReporter` where you used to print information to `stdout` or `stderr` in JUnit 4.

This does not clarify the intend behind key value pairs, though, and since leaving them empty is an error, it would help to explain how to use them. Otherwise users might be under the impression that the reporter is only useful to output test data (like in `reportSeveralValues` in [the user guide](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dependency-injection)), but not status messages.

Or maybe add an overload that accepts a single string as a message.

### PR DESCRIPTION
**Pull request by nicolaiparlog**
Adds:

* `TestReporter::publishMessage`
* `ExtensionContext::publishReportMessage`
* a few tests/demos
* a gazillion text files to appease the console runner tests (better never change the output format)

Fixes #1515.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.3.0-RC1.adoc [hunk 1]
```diff
@@ -87,6 +87,7 @@ on GitHub.
   class name and system hash code, separated by an `@` symbol.
 * New `getAs<Class>(index)` Kotlin extension method to make `ArgumentsAccessor` friendlier
   to use from Kotlin.
+* New method `publishMessage(String)` on `TestReporter` makes it easier to publish status messages without having to come up with a key (as is required by `publishEntry`).
```
Comment (sormuras) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.3.0-RC1.adoc]: Line too long. Please break at around 80-90 chars.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/TestReporter.java [hunk 2]
```diff
@@ -52,9 +53,22 @@
...
+	/**
+	 * Publish the specified message to be consumed by an
+	 * {@code org.junit.platform.engine.EngineExecutionListener}.
+	 *
+	 * @param message the message to be published; never {@code null} or blank
+	 * @see #publishEntry(Map)
+	 * @see #publishEntry(String, String)
+	 */
+	default void publishMessage(String message) {
+		this.publishEntry("message", message);
```
[the remark this answers is not in the document]
Comment (sormuras) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/TestReporter.java]: I wouldn't make it a constant -- but document that this convenience overload uses `message` as the `key` for publishing a report entry.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ExtensionContext.java [hunk 3]
```diff
@@ -274,6 +274,16 @@ default void publishReportEntry(String key, String value) {
 		this.publishReportEntry(Collections.singletonMap(key, value));
 	}
 
+	/**
+	 * Publish the specified message to be consumed by an
+	 * {@code org.junit.platform.engine.EngineExecutionListener}.
+	 *
```
Comment (sormuras) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/ExtensionContext.java]: Document that this convenience overload uses `message` as the `key` for publishing a report entry here too.

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/ReportingTests.java [hunk 4]
```diff
@@ -31,7 +31,7 @@
 class ReportingTests extends AbstractJupiterTestEngineTests {
 
 	@Test
-	void threeReportEntriesArePublished() {
+	void reportEntriesArePublished() {
```
Comment (sormuras) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/ReportingTests.java]: `three` was wrong all along...

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/TestReporter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nicolaiparlog) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/TestReporter.java]: Is that settled or will there be other opinions? Speak today or forever hold your peace. :wink:

**Comment by sbrannen** [comment]
Tentatively slated for 5.3 RC1 since this PR is relatively close to completion.

**Comment by nicolaiparlog** [comment]
Assuming the build passes, this should be ready to merge.

**Comment by sormuras** [comment]
Thanks @nicolaiparlog

**Comment by sbrannen** [comment]
Yep, thanks @nicolaiparlog! 👍

**Comment by sbrannen** [comment]
FYI: the following TODO had not actually been completed.

`Public API has Javadoc and @API annotations`

The `@API` and `@since` declarations were missing. 😈

**Comment by sbrannen** [comment]
But fret not, I added those in 23b167a08e144888adc21e94cd767199f4f78fbe. 😉

**Comment by sbrannen** [comment]
I'm sure it _could_ be automated. We discussed it once a long time ago, but (IIRC) we decided not to do it since it would require some custom coding for various tools (e.g., Checkstyle). I don't recall the details but rather only that we figured it wasn't worth the effort.

However, if you know of a way to achieve it with little effort (or effort you're willing to put forth), by all means let us know.

**Pull request merged**